RNA_seq
=======

Scripts related to RNA sequencing, transcriptomics, and analysis of differential gene expression

---

plotSeqLengthDist.R

This makes a plot of sequence length distributions, comparing multiple files, initially built to compare the output from different de novo assemblies of transcriptomes.  See the script header for usage.

---

xprs_2_edgeR.R

This is designed to do a series of analyses to explore differential gene expression from an RNA-seq dataset, particularly RNA-seq data from non-model organisms with only a de-novo assembly.  It needs two kinds of files as input: 

1) a set of xprs.results files from eXpress, one from each library (http://bio.math.berkeley.edu/eXpress/overview.html) 

2) an annotation file from Trinotate (http://trinotate.sourceforge.net/).  

See the EdgeR documentation for more details:
http://www.bioconductor.org/packages/2.13/bioc/vignettes/edgeR/inst/doc/edgeRUsersGuide.pdf

This is currently meant to be used interactively in an IDE such as R Studio.

--

GSEA_w_topGO.R

This takes the topTags*.csv files generated by xprs_2_edgeR.R and performs a gene set enrichment analysis (GSEA) based on the GO terms from a Trinotate annotation file.  It implements the Fisher exact test on the tags found to be significantly differentially expressed in xprs_2_edgeR.R but can be adapted for other tests; see the topGO documentation:
http://www.bioconductor.org/packages/release/bioc/vignettes/topGO/inst/doc/topGO.pdf

This is currently meant to be used interactively in an IDE such as R Studio.
