RNA_seq
=======

Scripts related to RNA sequencing, transcriptomics, and analysis of differential gene expression

Contains two directories:
diff_expression/
transcriptome_analysis/

--

transcriptome_analysis/plotSeqLengthDist.R

This makes a plot of sequence length distributions, comparing multiple files, initially built to compare the output from different de novo assemblies of transcriptomes.  See the script header for usage.

--

diff_expression/get_express_data.R 

Gets eXpress expression estimate data for subequent use in R.

--

diff_expression/get_DEG_list.R

Generates a list of differentially expressed genes, using the R object created by get_express_data.R  This is currently meant to be used interactively in an IDE such as R Studio.

--

diff_expression/GSEA_w_topGO.R

This takes the topTags*.csv files generated by xprs_2_edgeR.R and performs a gene set enrichment analysis (GSEA) based on the GO terms from a Trinotate annotation file.  It implements the Fisher exact test on the tags found to be significantly differentially expressed in xprs_2_edgeR.R but can be adapted for other tests; see the topGO documentation:
http://www.bioconductor.org/packages/release/bioc/vignettes/topGO/inst/doc/topGO.pdf

This is currently meant to be used interactively in an IDE such as R Studio.
