
# Scott Fay 
# 21 Mar 2014

#####
#
# Generates a list of differentially expressed genes based on files generated by "get_express_data.R"
#
# USAGE: Interactive, in a IDE like RStudio, because the decisions about what model to use and what factors to test against are complex
#
#####

# to install packages:
#install.packages("ggplot2")
#source("http://bioconductor.org/biocLite.R")
# biocLite()
#biocLite("edgeR")

library(ggplot2)
library(edgeR)
library(RColorBrewer)

#choose the output "diff_expression_data.RData" from get_express_data.R here to get the file path
getfilepath = file.choose()

# set the working directory with the output from get_express_data.R
setwd(dirname(getfilepath))

# load FPKM values, counts and groupings (output of import_express_data.R script)
load(getfilepath)
group

# get annotation file, "trinotate_annotation_report.txt", excel output from trinotate, exported from excel as tab-delimited text
transcripts <- read.delim(file.choose())

##########
#
# if you want to remove any data from the analyses, here's how to do it!
# 
##########

### as a test, remove two samples that are in columns 6 and 9
new_count <- all_count[,-c(6,9)]
all_count <- new_count
head(all_count)  # columns you removed should be gone
# also need to remove the corresponding rows from the group object
new_group <- group[-c(6,9),] 
head(new_group) # confirm rows are gone
group <- new_group
# and from the all_FPKM file
new_fpkm <- all_fpkm[,-c(6,9)]
head(new_fpkm) # confirm columns are gone
all_fpkm <- new_fpkm
#remove temporary object
rm(new_count, new_group, new_fpkm)
#save new Rdata file with new name
save(all_count, all_fpkm, group, file = "diff_expression_data_edited.RData")

##########
#
# EdgeR: Set up differential gene expression (DGE) analysis, generate summary plots:
#   define a model matrix *
#   estimate dispersion
#   make a plot of biological coefficient of variation vs. log(CPM)
#   make a multidimensional scaling plot of total counts
#   fit GLM for each transcript
#     * Note, how to define the statistical model is completely dependent on your experimental design and what question(s) you want to address.  See the EdgeR documentation as a first-pass guide: http://www.bioconductor.org/packages/2.12/bioc/vignettes/edgeR/inst/doc/edgeRUsersGuide.pdf
#
##########

# DGEList makes an EdgeR object using the column that defines the grouping (in this case "temp" for temperature)
y <- DGEList(counts=all_count, group=group$temp)
# filter out transcripts that do not have at least two reads out of 1,000,000 reads mapped in at least 4 samples
y <- y[rowSums(1e+06 * y$counts/expandAsMatrix(y$samples$lib.size, dim(y)) > 2) >= 4, ]
# reset the library sizes after filtering
y$samples$lib.size <- colSums(y$counts)
# TMM normalization compensates not just for library size but also the relative expression level among transcripts
y <- calcNormFactors(y)
# estimate dispersion
y <- estimateCommonDisp(y) 
y <- estimateTagwiseDisp(y)
# plot genewise biological coefficient of variation against gene abundance
plotBCV(y)
# make plot as a pdf
pdf(file="BCV_plot.pdf", height=6, width=6)
plotBCV(y, main = "Biological Coefficient of Variation")
dev.off()
# MDS plot
plotMDS(y , main = "MDS Plot for Count Data", labels = colnames( y$counts ), cex=0.7)
# make a pdf
pdf(file="MDS_plot.pdf", height=6, width=6)
plotMDS(y , main = "MDS Plot for Count Data", labels = colnames( y$counts ), cex=0.7)
plotMDS(y , main = "MDS Plot for Count Data", labels = group, cex=0.7)
dev.off()

# Plot total number of mapped reads to each transcript: look for outliers of library size
boxplot(log2(all_count+1), las=2) 
pdf(file="library_size_box_plot.pdf", height=7, width=10)
boxplot(log2(all_count+1), las=2) 
dev.off()

#####
#
# Find significantly differentially expressed genes between comparisons
#
#####

#####
# Function: get_DEGs
# perform LRT (likelihood ratio test) on certain factors
# returns a data frame that contains both toptags and annotation data
# arguments:
#   lrt = glmLRT object of the comparison you're making
#   annot = transcripts annotation data frame
#   fdr = false discovery rate (default 0.05)
#   critFC = a threshold fold change (default 2-fold)
#   onlyAnnot = flag whether to save only annotated transcripts
#
# saves files:
#   .pdf of the MA plot
#   .csv file of topTags, filtered by critical FC and onlyAnnot flag
#   
#####

# load the following function
get_DEGs <- function(lrt, annot, fdr=0.05, critFC=2, onlyAnnot=FALSE) {
  DEG <- summary(decideTestsDGE(lrt, p=fdr, adjust="BH")) # gives numbers of genes up and downregulated at FDR < pval
  Num_DEG <- (DEG[1] + DEG[-1])[2] # gets total number of DEGs based on the FDR
  tTags <- topTags(lrt, n=Num_DEG) # gets a list of DEGs, "topTags"
  comp <- paste(DEGs$comparison[1], "v", DEGs$comparison[2], sep="")
  cat("Comparison:", comp, "\n")
  cat("Total number of genes:", nrow(lrt$table), "\n")
  cat("Number of differentially expressed genes with FDR <", fdr, "=", nrow(tTags), "\n")
  # write an MA Plot .pdf
  cat("...saving an MAplot:", paste("MAplot_", comp, "_FDR_", fdr, ".pdf", sep=""), "\n" )
  pdf(file=paste("MAplot_", comp, "_FDR_", fdr, ".pdf", sep=""), height=6, width=6)
  detags <- rownames(tTags$table) # n= has to be the number of significantly differentially expressed genes
  plotSmear(lrt,de.tags=detags, cex=0.5) # plot of fold change given CPM, red for those < FDR
  abline(h=c(-1,1), col="dodgerblue") # blue line at twofold change
  dev.off()
  # Merge annotations with DGE stats
  tTags_frame <- tTags$table # make data frame from tTags
  tTags_frame$transcript_id <- rownames(tTags_frame)
  join <- merge(tTags_frame, annot) # gets the intersection of detags and transcripts, i.e., the annotations for the DE transcripts
  # Some summary stats
  cat("percent all transcripts with ORFs:", sum(annot$prot_id != ".") * 100 / nrow(annot), "\n")
  cat("percent toptags with ORFs:", sum(join$prot_id != ".") * 100 / nrow(join), "\n")
  cat("percent all transcripts with Pfam or BlastX or BlastP annotation:", sum((annot$Pfam != ".") | (annot$Top_BLASTX_hit != ".") | (annot$Top_BLASTP_hit != ".")) * 100 / nrow(annot), "\n")
  cat("percent toptags with Pfam or BlastX or BlastP annotation:", sum((join$Pfam != ".") | (join$Top_BLASTP_hit != ".") | (join$Top_BLASTX_hit != ".")) * 100 / nrow(join), "\n")
  cat("number of genes upregulated, (FC > ", critFC, "):", sum(join$logFC > log2(critFC)), "\n")
  cat("number of genes downregulated, (FC < ", 1/critFC, "):", sum(join$logFC < -log2(critFC)), "\n")
  join <- join[ abs(join$logFC) > log2(critFC) , ]
  # write a csv and return the relevant dataframe
  if(onlyAnnot) {
    cat("saving .csv of up or downregulated genes, given critical FC, with only annotated sequences:", paste("topTags_", comp, "_", "critFC", critFC, "_", fdr, "FDR_only_w_Annot.csv", sep="" ))
    write.csv( join[ (join$Pfam != ".") | (join$Top_BLASTX_hit != ".") | (join$Top_BLASTP_hit != ".") , ], file=paste("topTags_", comp, "_", "critFC", critFC, "_", fdr, "FDR_only_w_Annot.csv", sep="" ) )
    return(join[ (join$Pfam != ".") | (join$Top_BLASTX_hit != ".") | (join$Top_BLASTP_hit != ".") , ])
  } else {
    cat("saving .csv of up/downregulated genes, given a critical FC:", paste("topTags_", comp, "_", "critFC", critFC, "_", fdr, "FDR.csv", sep="" ))
    write.csv( join, file=paste("topTags_", comp, "_", "critFC", critFC, "_", fdr, "FDR.csv", sep="" ) )
    return(join)    
  }
}

#######
#
#  use exactTest for pairwise comparisons
#
#######

# for example, between temperatures 15 and 30
et <- exactTest(y, pair=(c("15", "30")))
summary(decideTestsDGE(et, p=0.01, adjust="BH"))

DEGs <- topTags(et, n=100)
DEGs$comparison

DEGlist_Di_15vs30_FC1 <- get_DEGs(et, annot=transcripts, fdr=0.000001, critFC=2)

# for all the possible pairwise comparisons, you'd want to rerun the above lines for each pair
# 15 vs 20
DEGlist_Di_15vs20_FC1 <- get_DEGs(exactTest(y, pair=(c("15", "20"))), annot=transcripts, fdr=0.000001, critFC=2)

# 15 vs 25
DEGlist_Di_15vs25_FC1 <- get_DEGs(exactTest(y, pair=(c("15", "25"))), annot=transcripts, fdr=0.000001, critFC=2)

# 25 vs 30
DEGlist_Di_25vs30_FC1 <- get_DEGs(exactTest(y, pair=(c("25", "30"))), annot=transcripts, fdr=0.000001, critFC=2)

# 20 vs 25
DEGlist_Di_20vs25_FC1 <- get_DEGs(exactTest(y, pair=(c("20", "25"))), annot=transcripts, fdr=0.000001, critFC=2)

# 20 vs 30
DEGlist_Di_20vs30_FC1 <- get_DEGs(exactTest(y, pair=(c("20", "30"))), annot=transcripts, fdr=0.000001, critFC=2)



# make a merged list of contigs with fold change of 4
DEGlist_Di_15vs30_FC4 <- get_DEGs(exactTest(y, pair=(c("15", "30"))), annot=transcripts, fdr=0.000001, critFC=4)

# for all the possible pairwise comparisons, you'd want to rerun the above lines for each pair
# 15 vs 20
DEGlist_Di_15vs20_FC4 <- get_DEGs(exactTest(y, pair=(c("15", "20"))), annot=transcripts, fdr=0.000001, critFC=4)

# 15 vs 25
DEGlist_Di_15vs25_FC4 <- get_DEGs(exactTest(y, pair=(c("15", "25"))), annot=transcripts, fdr=0.000001, critFC=4)

# 25 vs 30
DEGlist_Di_25vs30_FC4 <- get_DEGs(exactTest(y, pair=(c("25", "30"))), annot=transcripts, fdr=0.000001, critFC=4)

# 20 vs 25
DEGlist_Di_20vs25_FC4 <- get_DEGs(exactTest(y, pair=(c("20", "25"))), annot=transcripts, fdr=0.000001, critFC=4)

# 20 vs 30
DEGlist_Di_20vs30_FC4 <- get_DEGs(exactTest(y, pair=(c("20", "30"))), annot=transcripts, fdr=0.000001, critFC=4)

merged_table = rbind(DEGlist_Di_15vs30_FC4,DEGlist_Di_15vs20_FC4,DEGlist_Di_15vs25_FC4,DEGlist_Di_25vs30_FC4,DEGlist_Di_20vs25_FC4,DEGlist_Di_20vs30_FC4)

head(merged_table)
merged_table_annotation= merged_table[,-c(2:5)]
head(merged_table_annotation)
merged_table_unique = unique(merged_table_annotation)
head(merged_table_unique)
sort(merged_table_unique)
write.csv(merged_table_unique,"merged_table_unique.csv")
